### PR TITLE
Add card buying

### PR DESCRIPTION
**Pull request by oliverlee**

### CODE REVIEW

File: src/dominion/player.rs [hunk 1]
```diff
@@ -191,10 +234,81 @@ impl Player {
...
+
+        p.phase = Some(TurnPhase::Buy {
+            remaining_buys: 1,
+            total_wealth: 0,
+        });
+
+        let r = p.buy_card(&CardKind::Copper);
+
+        assert!(r.is_ok());
+        assert_eq!(p.discard_pile[0], &CardKind::Copper);
```
Comment (andebjor) [inline src/dominion/player.rs:266]: Assert that the supply of coppers have decreased by one?

File: src/dominion/player.rs [hunk 2]
```diff
@@ -191,10 +234,81 @@ impl Player {
...
+
+        p.phase = Some(TurnPhase::Buy {
+            remaining_buys: 1,
+            total_wealth: 5,
+        });
+
+        let r = p.buy_card(&CardKind::Market);
+
+        assert!(r.is_ok());
+        assert_eq!(p.discard_pile[0], &CardKind::Market);
```
Comment (andebjor) [inline src/dominion/player.rs:289]: Assert that wealth is now 0?

File: src/dominion/player.rs [hunk 3]
```diff
@@ -136,6 +142,43 @@ impl Player {
         }
     }
 
+    pub fn buy_card(&mut self, c: &'static CardKind) -> Result<()> {
```
Comment (andebjor) [inline src/dominion/player.rs]: I'm not a fan of single letter arguments, such as `c` :)

**Comment by oliverlee** [comment]
#![feature(vec_remove_item)] only works on nightly. Should we drop support for stable and beta?

File: src/dominion/player.rs (hunk 1 above)
Comment (oliverlee) [reply src/dominion/player.rs:266]: done

**Comment by mickvangelderen** [comment]
> #![feature(vec_remove_item)] only works on nightly. Should we drop support for stable and beta?

For a toy project yes, we can worry about running this in stable later (read: probably never).

**Review by mickvangelderen — COMMENTED** [review]
I am concerned about the general architecture. The simulation we are running concerns all of the game state (currently kept in Arena). Player hands are just a part of that. I think it is better to do all of the mutations in a function which has access to the entire state. This will remove the need for reference counted supply pointers (and probably more in the future if this design is continued).

I am okay with going forward with this architecture for the sake of learning though.

**Comment by oliverlee** [comment]
> I am concerned about the general architecture. The simulation we are running concerns all of the game state (currently kept in Arena). Player hands are just a part of that. I think it is better to do all of the mutations in a function which has access to the entire state. This will remove the need for reference counted supply pointers (and probably more in the future if this design is continued).
> 
> I am okay with going forward with this architecture for the sake of learning though.

This is mostly a learning exercise but I think we should change the architecture if we can come up with a better idea. Resolving a stack of actions as a result of playing a card will probably be challenging with this approach.

**Pull request merged**